### ISSUE TITLE
How to fintune a pretrained network? 

### ISSUE THREAD

**Issue by camel007** [facebookarchive/caffe2#868]
Is there some examples about this? Thank you

**Comment by cisi17** [comment]
Hello,

I am trying it too. I have thought some things:

-When you create your model:

train_model = cnn.CNNModelHelper(order, name="alexnet",init_params=False)

-After, I think that you can load the weights of the  network pretrained with your .pb file:

init_net= caffe2_pb2.NetDef()
with open(INIT_NET, 'r') as f:
        init_net.ParseFromString(f.read())

Then, you run the net:

workspace.RunNetOnce(init_net.SerializeToString())
workspace.CreateNet(train_model.net)

-Finally, you re-train your net.

I am thinking about this.

Please, if you know something, tell me.

thanks.

**Comment by cisi17** [comment]
Hello @camel007 ,

I am trying it too. I have thought some things:

-When you create your model:

train_model = cnn.CNNModelHelper(order, name="alexnet",init_params=False)

-After, I think that you can load the weights of the network pretrained with your .pb file:

init_net= caffe2_pb2.NetDef()
with open(INIT_NET, 'r') as f:
init_net.ParseFromString(f.read())

Then, you run the net:

workspace.RunNetOnce(init_net.SerializeToString())
workspace.CreateNet(train_model.net)

-Finally, you re-train your net.

I am thinking about this.

Please, if you know something, tell me.

thanks.

**Comment by shankarapailoor** [comment]
Could you explain what you mean by finetune? Do you mean reinitializing weights or changing the shape of layers?

**Comment by camel007** [comment]
Thanks a lot1，initialize with exist model2，clone part of the origin network and append new layer发自我的iPhone------------------ Original ------------------From: shankarapailoor  <notifications@github.com>Date: 周一,6月 26,2017 23:34To: caffe2/caffe2 <caffe2@noreply.github.com>Cc: 张垚 <yaozhang.chn@gmail.com>, Mention <mention@noreply.github.com>Subject: Re: [caffe2/caffe2] How to fintune a pretrained network? (#868)Could you explain what you mean by finetune? Do you mean reinitializing weights or changing the shape of layers?

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/caffe2/caffe2","title":"caffe2/caffe2","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/caffe2/caffe2"}},"updates":{"snippets":[{"icon":"PERSON","message":"@shankarapailoor in #868: Could you explain what you mean by finetune? Do you mean reinitializing weights or changing the shape of layers?"}],"action":{"name":"View Issue","url":"https://github.com/caffe2/caffe2/issues/868#issuecomment-311095434"}}}

**Comment by camel007** [comment]
Thanks a lot
I want to reinitializing the weights and change some layer

**Comment by shankarapailoor** [comment]
The way I have done this so far is by modifying the protobuf definition of the model. You can do this in python as follows:

I can provide an example. Let's say I want to change the last Fully Connected Layer of AlexNet to classify Food-101 instead of ImageNet. That means I want the last fully connected layer fc8 to have 101 outputs instead of 1000.  The way I did this is by changing the dimensions of fc8_w, and fc8_b in the param_init_net of the model. In the model zoo for Alex Net you can find the weights for fc8 under the op with output "fc8_w" and the biases under the op with output "fc8_b"

```python
for op_ in train_model.param_init_net.Proto().op:
     for output in op_.output:
            if output == "fc8_w":
                  for arg_ in op.arg:
                          if arg_.name == 'shape':
                                arg.ClearField('ints')
                                arg_.ints.extend([101, 1024])
                          elif arg_.name == 'values':
                                arg.ClearField('floats')
                                arg_.floats.extend(np.random.normal(0, .1, 1024*1000))
```
To add a parameter to gradient descent and initialize the weights you can do that as follows:

```python
weight_initializer = intializers.update_initializer(None, None, ("XavierFill", {}))
model.create_param(param_name='fc8_w',  shape=[101, 1024], initializer=weight_initializer, tags=ParameterTags.WEIGHT)

```

**Comment by camel007** [comment]
@shankarapailoor 
Appreciate for your reply

**Comment by marcelogarcia94** [comment]
Hi @shankarapailoor,

thanks for your help in my project. But I have a problem. When I run:

order="NCHW"
train_model = cnn.CNNModelHelper(order, name="net")
train_model.param_init_net.AppendNet(core.Net(init_net))
train_model.net.AppendNet(core.Net(predict_net))

for op_ in train_model.param_init_net.Proto():
     for output in op_.output:
            if output == "fc8_w":
                  for arg_ in op.arg:
                          if arg_.name == 'shape':
                                arg.ClearField('int')
                                arg_.ints.extend([101, 1024])
                          elif arg_.name == 'values':
                                arg.ClearField('floats')
                                arg_.floats.extend(np.random.normal(0, .1, 1024*1000))

An error appear:

'NetDef' object is not iterable

Do you know what is the reason?

Thank you.

**Comment by shankarapailoor** [comment]
Sorry, my mistake. I think you need to do:

```python
for op_ in train_model.param_init_net.Proto().op:
```

You also need to change arg.ClearField('int') to arg.ClearField('ints')
The protobuf definition is over here: https://github.com/caffe2/caffe2/blob/master/caffe2/proto/caffe2.proto. The NetDef is defined on line 166. As you can see the operators are a field named "op" and it is repeated so you can iterate over op.

**Comment by marcelogarcia94** [comment]
Hello @shankarapailoor ,

I think that there are some small mistakes:

for op_ in train_model.param_init_net.Proto().op:
     for output in op_.output:
            if output == "fc8_w":
                  for arg_ in op.arg:
                          if arg_.name == 'shape':
                                **arg.ClearField('ints')
                                arg_.ints.extend([101, 1024])
                          elif arg_.name == 'values':
                                **arg.ClearField('floats')
                                arg_.floats.extend(np.random.normal(0, .1, 1024*1000))

In **, I think that you have to put arg_ instead of arg. It is a small mistake.

But when I run the code, I new error appear in line (arg_.floats.extend(np.random.normal(0, .1, 1024*1000))):

The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

If I try to put arg_.floats.extend([np.random.normal(0, .1, 1024*1000)]), using []. Another error appear:

array([-0.01383945,  0.20971348,  0.07296314, ...,  0.04011999,
       -0.05002904,  0.01741383]) has type <type 'numpy.ndarray'>, but expected one of: (<type 'float'>, <type 'int'>, <type 'long'>)

I think that is a small error. Do you know something?


I wanted to ask you another thing if you don't matter. How do you crate the new network with the other weights? How do you pass the data?

Thanks you so much.

**Comment by attitudechunfeng** [comment]
Some tips for u, I first save the params with pickle when training the model, and when I want to retrain the model, the only thing i need do is to load the saved params from pickle and replace with the loaded params. Though it seems ugly, it worked for me.

**Comment by camel007** [comment]
Thanks a lot @attitudechunfeng

**Comment by CriCL** [comment]
hi @camel007, where you able to retrain AlexNet?

I need some help!
